### PR TITLE
Adding compose()

### PR DESCRIPTION
**Pull request by mschwager**
Hey!

Since it looks like https://github.com/lstrojny/functional-php/pull/74 hasn't been making any progress I've decided to submit my own PR.

Let me know if you have any concerns :)

### CODE REVIEW

File: src/Functional/Compose.php [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+use function Functional\id;
+
+/**
+ * Return a new function that composes all functions in $functions into a single callable
+ *
+ * @param array $functions
+ * @return callable
+ */
+function compose(...$functions)
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Can be `callable ...$functions`

File: src/Functional/Compose.php [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+ * THE SOFTWARE.
+ */
+namespace Functional;
+
+use function Functional\id;
+
+/**
+ * Return a new function that composes all functions in $functions into a single callable
+ *
+ * @param array $functions
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Should be `callable[] $functions`

File: src/Functional/Compose.php [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+function compose(...$functions)
+{
+    return array_reduce(
+        $functions,
+        function ($carry, $item) {
+            return function ($x) use ($carry, $item) {
+                return $item($carry($x));
+            };
+        },
+        "Functional\id"
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Single quotes please

File: src/Functional/Compose.php [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ * @param array $functions
+ * @return callable
+ */
+function compose(...$functions)
+{
+    return array_reduce(
+        $functions,
+        function ($carry, $item) {
+            return function ($x) use ($carry, $item) {
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Shouldn’t the inner function have n-arity as well?

**Comment by lstrojny** [comment]
Thanks for picking that one up, looks good already, just a couple of small questions inline.

File: src/Functional/Compose.php (hunk 4 above)
Comment (mschwager) [inline src/Functional/Compose.php]: Hmm, conceptually this makes sense to me. However, there is a problem. The identity function doesn't have n-arity, and making one that does is difficult (impossible?). For example:

```php
>>> $id = function (...$args) { return $args; };
>>> $id(1, 2, 3);
=> [
     1,
     2,
     3,
   ]
```

We get a boxed result, despite passing in an unboxed set of arguments, so it's not a true identity function.

Thoughts?

**Comment by mschwager** [comment]
Info on the Travis-CI failure here: https://github.com/facebook/hhvm/issues/4858

File: src/Functional/Compose.php (hunk 4 above)
Comment (lstrojny) [inline src/Functional/Compose.php]: Hmm, interesting problem. We could get rid of calling `id()` and checking if `$item` is != null.

Comment (mschwager) [inline src/Functional/Compose.php]: Let me know what you think of this approach.

File: src/Functional/Compose.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+ * @return callable
+ */
+function compose(...$functions)
+{
+    return array_reduce(
+        $functions,
+        function ($carry, $item) {
+            return function (...$x) use ($carry, $item) {
+                $result = $carry(...$x);
+                if (is_array($result)) {
```
Comment (lstrojny) [inline src/Functional/Compose.php]: Unboxing all arguments all of the time is something that I wouldn’t do, as that will lead to opaque behavior.

Comment (mschwager) [inline src/Functional/Compose.php]: Yeah, I was thinking about this last night. It's going to be difficult to distinguish between actually returning an `array` and returning an `array` which should be unboxed without some additional plumbing.

What do you think of taking the original approach? That is, composing functions that only take one argument.

Comment (lstrojny) [inline src/Functional/Compose.php]: What do you think about this:

```php
function compose(callable ...$functions) {
    return function(...$arguments) use ($functions) {
        return array_reduce(
            $functions,
            function($value, callable $function) {
                return $function($value);
            },
            ...$arguments
        );
    };
}
```

Comment (lstrojny) [inline src/Functional/Compose.php]: Ignore it, it’s wrong :hand:

Comment (lstrojny) [inline src/Functional/Compose.php]: After thinking about it a bit more, let’s go with a 1-arity compose, I don’t think it’s worth the hazzle

File: src/Functional/Compose.php [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+use function Functional\id;
+
+/**
+ * Return a new function that composes all functions in $functions into a single callable
+ *
+ * @param callable[] $functions
+ * @return callable
+ */
+function compose(callable ...$functions)
```
Comment (mschwager) [inline src/Functional/Compose.php]: Not sure what's more important to you, having the type hinting, or the build passing on HHVM.

Comment (lstrojny) [inline src/Functional/Compose.php]: We can just have two variants. Have a look at `partial_any()`

**Comment by mschwager** [comment]
So, there's an issue with the Travis build and HHVM. I've added a comment on the issue here: https://github.com/facebook/hhvm/issues/4858#issuecomment-197621161

Adding a condition around `defined('HHVM_VERSION')` isn't sufficient since the linting process still checks all the code. It's not conditionally included based upon `define`'s. There's also no HHVM `FIXME` to squelch this error (as seen in the comment above).

We have a couple options to move this forward. Allow failures on HHVM (not sure how many users of this library are on HHVM) with something like `.travis.yml`:

```
...
matrix:
  allow_failures:
    - php: hhvm
...
```

Or avoid the type hint all together. It's a nice-to-have, but I wouldn't say it's critical to this code.

Let me know what you think.

**Comment by lstrojny** [comment]
Yeah, let’s do the allow_failures option for now. I hope hhvm will soon fix the type hint issue anyway.

**Comment by mschwager** [comment]
Not sure what's going on here with Scrutinizer. Any ideas?

**Comment by lstrojny** [comment]
Nope, but I don’t care for now.

**Comment by lstrojny** [comment]
Thank you!

**Pull request merged**